test-data/unit/check-dataclass-transform.test

```diff
@@ -202,3 +209,42 @@ Foo(5)
 
 [typing fixtures/typing-full.pyi]
 [builtins fixtures/dataclasses.pyi]
+
+[case testDataclassTransformViaBaseClass]
+# flags: --python-version 3.11
+from typing import dataclass_transform
+
+@dataclass_transform(frozen_default=True)
+class Dataclass:
+    def __init_subclass__(cls, *, kw_only: bool): ...
+
+class Person(Dataclass, kw_only=True):
+    name: str
+    age: int
+
+reveal_type(Person)  # N: Revealed type is "def (*, name: builtins.str, age: builtins.int) -> __main__.Person"
+Person('Jonh', 21)  # E: Too many positional arguments for "Person"
+person = Person(name='John', age=32)
+person.name = "John Smith"  # E: Property "name" defined in "Person" is read-only
+
+[typing fixtures/typing-full.pyi]
+[builtins fixtures/dataclasses.pyi]
+
+[case testDataclassTransformViaMetaclass]
+# flags: --python-version 3.11
+from typing import dataclass_transform
+
+@dataclass_transform(frozen_default=True)
+class Dataclass(type): ...
+
+class Person(metaclass=Dataclass, kw_only=True):
+    name: str
+    age: int
+
+reveal_type(Person)  # N: Revealed type is "def (*, name: builtins.str, age: builtins.int) -> __main__.Person"
+Person('Jonh', 21)  # E: Too many positional arguments for "Person"
+person = Person(name='John', age=32)
+person.name = "John Smith"  # E: Property "name" defined in "Person" is read-only
+
+[typing fixtures/typing-full.pyi]
+[builtins fixtures/dataclasses.pyi]
```

Comment:
Semi-related comment: Can you add some incremental mode test cases as well to test serialization (no need to do it in this PR)? Also other dataclass features could be covered. Similarly, daemon test cases would be nice, but again it can happen later. I can help if you are not sure how these work.

Reply:
Can do.

> Also other dataclass features could be covered.

Could you clarify this a bit? I was trying to avoid covering _all_ features since that would duplicate the a lot of test cases, but I'm happy to expand the test file.

Reply:
It would be good to test serialization of all the new attributes in incremental mode test cases.

To test mypy daemon properly, something like this could a good starting point:
* Test propagating changes after changing flags (e.g. frozen).
* Test switching between regular class and dataclass transform class.
* If a base class or meta class is turned into a dataclass transform, the change should be propagated to other classes that use them.
* Test modifying the signature of an attribute. This should propagate changes to subclasses in different modules that inherit that class.

The daemon test cases may need code changes, so it may be best to add them in a separate PR.

Reply:
It may be best to first add one or two daemon test cases, and once we've established a good way to write them, adding the remaining ones should be easy.